libsel4camkes/docs/virtqueue.md

```diff
@@ -0,0 +1,133 @@
+<!--
+     Copyright 2021, Data61, CSIRO (ABN 41 687 119 230)
+
+     SPDX-License-Identifier: BSD-2-Clause
+-->
+
+# Virtqueues
+
+libsel4camkes provides an implementation of the virtqueue transport mechanism.
+virtqueues can be used to asynchronously transfer data between two components.
+Although virtqueues are mostly used between clients and device drivers, this
+transport mechanism is useful for general producer-consumer patterns.
+
+## Usage
+
+The virtqueue libraries requires components to first initialise a
+`seL4VirtQueue` CAmkES connection between components. Here's an example of how
+this is done:
+
+```c
+component Foo {
+    uses VirtQueueDrv drv;
+}
+
+component Foo2 {
+    uses VirtQueueDev dev;
+}
+
+assembly {
+    composition {
+        component VirtQueueInit vq_init;
+        component Foo bar;
+        component Foo2 baz;
+        
+        connection virtqueue_foo seL4VirtQueue(to vq_init.init, from bar.drv, from baz.dev);
+    }
+    
+    configuration {
+        bar.drv_id = 1;
+        baz.dev_id = 1;
+        bar.drv_shmem_size = 4096;
+        baz.dev_shmem_size = 4096;
+        virtqueue_foo.init_topology = [{"drv" : "bar.drv", "dev": "baz.dev" }];
+    }
+}
+```
+
+From the example above, the driver side of the virtqueue is `bar` and the
+device side is `baz`. The IDs of this particular virtqueue channel is 1 for
+both components. It is possible to have multiple channels for a component, but
+each channel's ID must be different. The memory size of the channel is 4096 and
+can also be changed, however, it must be page-aligned to 4096.
```

Comment:
Shouldn't this be dependent on the page-size of the hardware?  Or are pages *always* 4096 bytes in CAmkES systems, even if the underlying hardware uses a different base page size (ARM, x86, x86_64, aarch64 all use 4096 IA64 uses 64k, not sure about RISC-V)